app/models/practice_editor_session.rb

```diff
@@ -0,0 +1,129 @@
+class PracticeEditorSession < ApplicationRecord
+  belongs_to :practice
+
+   def self.practice_locked_for_current_user(cur_user_id, practice_id)
+    rec = PracticeEditorSession.where(practice_id: practice_id, session_end_time: nil).where.not(session_start_time: nil).order("session_start_time DESC").first()
+    return false if rec.blank?
+    if rec.user_id == cur_user_id
+      return false
+    end
+    return true
+  end
+  # returns 0 if not locked and returns the practice_editor_session id if locked
+  def self.practice_locked(practice_id)
+    rec = PracticeEditorSession.where(practice_id: practice_id, session_end_time: nil).where.not(session_start_time: nil).last
+    return 0 if rec.nil?
+    return rec.id
+  end
+
+  def self.session_out_of_time(rec_id)
+    rec = PracticeEditorSession.find_by_id(rec_id)
+    is_published = Practice.find_by_id(rec.practice_id).published
+    is_published ? session_end = DateTime.now.utc - 15.minutes : session_end = DateTime.now.utc - 30.minutes
+    if rec.session_start_time < session_end
+      rec.session_end_time = DateTime.now
+      rec.save
+      return true
+    end
+    return false
+  end
+
+  def self.lock_practice_for_user(user_id, practice_id)
+    PracticeEditorSession.create user_id: user_id, practice_id: practice_id, session_start_time: DateTime.now, session_end_time: nil
+  end
+  def self.locked_by(rec_id, include_time = true)
+    rec = PracticeEditorSession.find_by_id(rec_id)
```

Comment:
I think we should return the User here and then we can manipulate the text in a helper or the controller